### ISSUE TITLE
Improve baseof.html resolution

### ISSUE THREAD

**Issue by bep** [gohugoio/hugo#4893]
This is related to #4891 -- which due to how we do base template "decision making" (when we read the templates; we don't know enough to do a proper selection). 

I have noticed people complain about this, but have not thought too hard about it until I started to use it myself ...

One made-up example:

```bash
/layouts/magazine
baseof.html

/layouts/blog
list.html
single.html
```

If I in `/blog/_index.md` set `type: magazine` I really want to use the base template in ... `/magazine`. Now you will get the one in `_default` if present, because that is what we know when we parse these templates.

I think we can do bettter.

**Comment by regisphilibert** [comment]
I did not know about this behavior but it does go against the temptating logic I expect from Hugo (or any other frameworks’ template hierarchy). A lot of people must consider this as a bug and I wouldn’t blame them so I agree it should be fixed. 

> I don’t think this would be considered breaking. 

I think the number of project using more than one baseof is limited anyway. Some very rare users may have a surprise while updating so a warning would not hurt, but it is worth it.

**Comment by bep** [comment]
> A lot of people must consider this as a bug and I wouldn’t blame them so I agree it should be fixed.

It is not a bug. This is well defined in the documentation. The original implementation was obvious. Improving on that implementation is actually a fair amount of work, so he/she will certainly want the correct motivation of improving and not fixing.

I repeat: This is not bug fix. It is labelled correctly as an enhancement. Remember this for the future.

**Comment by regisphilibert** [comment]
Yes, just read the doc, I am definitely not educated enough on the matter.

So to make sure I understand this improvement correctly we would go from naming our magazine baseof `/layouts/_default/magazine-baseof.html` to the more logical `/layouts/magazine/baseof.html`?

Would make a lot of sense indeed.

**Comment by regisphilibert** [comment]
Sorry I read the [doc](https://gohugo.io/templates/base/#example-base-template-lookup-order) and found the behaviour described above in there (namely, `/themes/mytheme/layouts/post/baseof.html`) so I was a bit confused...

**Comment by bep** [comment]
If you find errors in documentation, please fix that. But what you refer to is:

```bash
/layouts/blog
baseof.html
list.html
single.html
```

Which works fine.

**Comment by budparr** [comment]
I agree this would be helpful. I tend to use block templates or conditionals to keep me from having too many baseof files and this would make them smarter.

**Comment by kaushalmodi** [comment]
Mine won't be a useful datapoint, I only use **one** `baseof.html` per site, in `layouts/_default/baseof.html`.

**Comment by bep** [comment]
This was fixed in 0.63.
